Написал код программы сумматора

St1ke opened the pull request:
1)Ввёл две переменные типа double, 
2)Дальше сделал так чтоб выводилось сообщение в консоли о том что нужно сделать, 3)Конвертировал обе переменных чтоб консоль могла считывать вводимые значения, 
4)Ввёл третью переменную которая обозначает сумму вводимых чисел
5)Написал код для вывода результата суммы в консоли

St1ke committed:
Написал код программы сумматора

1)Ввёл две переменные типа double, 
2)Дальше сделал так чтоб выводилось сообщение в консоли о том что нужно сделать, 3)Конвертировал обе переменных чтоб консоль могла считывать вводимые значения, 
4)Ввёл третью переменную которая обозначает сумму вводимых чисел
5)Написал код для вывода результата суммы в консоли

Tauders commented on task-2/Summator/Summator/Program.cs:
```diff
@@ -6,7 +6,14 @@ class Program
+            a = Convert.ToDouble(Console.ReadLine());
+            b = Convert.ToDouble(Console.ReadLine());
+            double total = a + b;
+            Console.WriteLine("Cумма " + a + " и " + b + " равна " + total + ".");
```
Используй интерполяцию строк.

St1ke committed:
Поменял сложение строк на интерполяцию

Tauders reviewed (changes requested):
Надо добавить обработку исключений

CodingFeles commented on task-2/Summator/Summator/Program.cs:
```diff
@@ -6,7 +6,14 @@ class Program
+            double b;
+
+            Console.WriteLine("Введите пару чисел для их сложения");
+            a = Convert.ToDouble(Console.ReadLine());
+            b = Convert.ToDouble(Console.ReadLine());
```
Про обработку исключений: здесь у тебя Console.ReadLine считывает строку, которую ввёл пользователь. Convert.ToDouble пытается превратить переданную ему строку в double. Что произойдёт если передать строку, которая не задержит числовых значений? Например: "abc" ?

St1ke committed:
[task-3 Summator] Добавил в код обработку исключения

Или показалось слишком простым заданием и конкретно тут смысл таков, что пользователь использующий данную программу получал информацию о том что именно сделал не так, или я не понял задачу. Я решил добавить вторую попытку для пользователя, знаю можно сделать более совершенно это циклами в будущем, но пока так.

Tauders commented on task-2/Summator/Summator/Program.cs:
```diff
@@ -6,7 +6,29 @@ class Program
+            double b;
+            try
+            {
+
```
Лишняя пустая строка

CodingFeles commented on task-2/Summator/Summator/Program.cs:
```diff
@@ -8,12 +8,27 @@ static void Main(string[] args)
+                double total = a + b;
+                Console.WriteLine($"Сумма {a} и {b} равна {total}.");
+            }
+            catch
```
Я не уверен, что стоит писать catch без каких-либо уточнений. Документация не рекомендует так делать. Однако, в данном случае не знаю - пусть @Tauders оценит.

Tauders replied:
Так делают, когда надо просто заглушить вообще все исключения, не сильно хорошая практика, и надо что-то делать с объектом исключения.

Давай сразу тогда и применим хорошую практику.

@St1ke добавь сюда код для получения исключения и выведи текст исключения на экран, когда говоришь пользователю, что он ввёл что-то не то.

Например:
```
Введены некоректные значения, пожалуйста, введите числа и попробуйте ещё раз. Полученная ошибка:
{Текст ошибки}
```

Текст ошибки на следующей строке

St1ke replied:
Не сразу понял о чем речь, тк в той главе которую @Tauders сказал прочитать для решения задачи ничего не было, но              когда изучал вопрос с помощью гугла натыкался на что то подобное, это типы исключений, я добавлю это, там вроде просто несколько категорий ошибок как я понял, но это нам не задавали), на метаните посмотрел, через одну главу как раз про типы есть.

CodingFeles replied:
Да, всё верно. По хорошему, catch должен "знать" какие типы исключений ловить, т.к. некоторые вещи лучше оставлять на откуп окружению.
"Нам не задавали" - к тебе никаких претензий нет. Ревью всегда будет критиковать работу по оформлению или существу, но в этом и смысл. Мы можем указывать на некоторые нюансы, даже если мы их не поясняли или они вне изучаемых тем. Если не получается самостоятельно разобраться или не очень понятно что делать - не стесняйся и обращайся к нам. Никаких негативных оценок тут нет.

St1ke replied:
> Да, всё верно. По хорошему, catch должен "знать" какие типы исключений ловить, т.к. некоторые вещи лучше оставлять на откуп окружению.
> "Нам не задавали" - к тебе никаких претензий нет. Ревью всегда будет критиковать работу по оформлению или существу, но в этом и смысл. Мы можем указывать на некоторые нюансы, даже если мы их не поясняли или они вне изучаемых тем. Если не получается самостоятельно разобраться или не очень понятно что делать - не стесняйся и обращайся к нам. Никаких негативных оценок тут нет.

Да просто пояснил в **рамках обучения** почему не знал, и понимаю почему и для чего делается.

Tauders replied:
@St1ke выведите текст ошибки ещё в консоль

Tauders commented on task-2/Summator/Summator/Program.cs:
```diff
@@ -6,7 +6,29 @@ class Program
+            }
+            catch
+            {
+                Console.WriteLine("Введите числовые значения.");
```
Здесь лучше писать, что было сделано не так - т.е., например, "Введены некоректные значения, пожалуйста, введите числа и попробуйте ещё раз".
В конкретно этом случае нормально и так, просто на будущее.

Tauders commented on task-2/Summator/Summator/Program.cs:
```diff
@@ -6,7 +6,29 @@ class Program
+            {
+                Console.WriteLine("Введите числовые значения.");
+            }
+            finally
+            {
+
+                a = Convert.ToDouble(Console.ReadLine());
+                b = Convert.ToDouble(Console.ReadLine());
+                double total = a + b;
+                Console.WriteLine($"Сумма {a} и {b} равна {total}.");
+            }
```
В текущем виде не нужен, т.к. что будет, если пользователь здесь введёт неправильные значения? Если хочешь, можешь использовать цикл бесконечный, чтобы пользователь мог вводить много раз, но в рамках задачи не требуется.

CodingFeles commented on task-2/Summator/Summator/Program.cs:
```diff
@@ -8,12 +8,27 @@ static void Main(string[] args)
+            {
+                Console.WriteLine("Введите числовые значения.");
+            }
+            finally
```
Блок finally нужен для вещей, который должны произойти независимо от того было брошено исключение в блоке try или нет.

А что будет, если тебе снова введут "abc"?
Тут два выхода:

1. Если ты хочешь дать возможность пользователю исправить свою ошибку и ввести что-то заного, то надо думать о том, как переделать программу, чтобы за один её запуск можно было проводить несколько сложений: тогда у тебя в случае исключения будет простой переход на следующую попытку сложения.
2. Если программа, как изначально была написана, предполагает только одну попытку сложить, то должна ли она заного предлагать пользователю исправиться? Или просто сообщить об ошибке и завершить свою работу.

CodingFeles replied:
Кстати, а ты заметил, что даже если ввёл правильные числа, то у тебя программа снова запросит числовые значения?

CodingFeles commented on task-2/Summator/Summator/Program.cs:
```diff
@@ -8,12 +8,27 @@ static void Main(string[] args)
+            finally
+            {
+
+                a = Convert.ToDouble(Console.ReadLine());
+                b = Convert.ToDouble(Console.ReadLine());
+                double total = a + b;
+                Console.WriteLine($"Сумма {a} и {b} равна {total}.");
```
Если ты хочешь вызывать один и тот же блок кода несколько раз - не копируй его. Для этого можно вынести код в отдельную функцию и уже её вызывать несколь раз.

St1ke replied:
> Если ты хочешь вызывать один и тот же блок кода несколько раз - не копируй его. Для этого можно вынести код в отдельную функцию и уже её вызывать несколь раз.

Да я знаю что циклами можно сделать всё красиво, как дойдем сделаю, хз просто решил сделать

CodingFeles replied:
Нет, про циклы - это не сюда. Здесь о том, что если в нескольких местах нужен один и тот же код, то скорее всего этот код надо вынести в отдельную функцию.

CodingFeles replied:
Про циклы: это несколькими комментариями выше, где обсуждения про то как реагировать на ошибку.

St1ke replied:
> Если ты хочешь вызывать один и тот же блок кода несколько раз - не копируй его. Для этого можно вынести код в отдельную функцию и уже её вызывать несколь раз.

я понял кажется о чем речь, или в джаве сталкивался или как в математике, что куча действия обозначаются чем то одним и коротким, но я в общем так не умею пока судя по всему, поэтому просто уберу самодеятельность

CodingFeles replied:
Да, вкратце функции это именованные блоки кода, у которых описаны данные которые они получают и какие возвращают. А ты можешь по имени вызывать их из других блоков кода.

Самодеятельность - это хорошо. Мы просто будем давать советы, как её сделать лучше.
Но в данном случае блок finally тут просто не нужен.

St1ke committed:
[task-2] Исправление ошибок в сумматоре

1)убрал строку
2)вроде? обозначил тип исключения в catch 
3)Сделал сообщение ошибки в catch верным
4)удалил блок finally

St1ke committed:
[task-2] Вывел текст ошибки в консоль

CodingFeles commented on task-2/Summator/Summator/Program.cs:
```diff
@@ -6,7 +6,20 @@ class Program
+                double total = a + b;
+                Console.WriteLine($"Сумма {a} и {b} равна {total}.");
+            }
+            catch (FormatException)
+            {
+                Console.WriteLine("Введены некорректные символы, пожалуйста попробуйте ещё раз используя числовые значения. Полученная ошибка: {FormatException}");
+            }
```
А ты проверял работает ли оно? Я не уверен, но сомневаюсь, что оно соберётся.

Просто FormatException - это тип. А интерполяция строк всё же ждёт, что ты будешь передавать какие-то значения или переменные.

Можно написать `catch (FormatException e)` и переменную `e` использовать в интерполяции. У исключений есть свойство Message, которое позволяет получить текст ошибки. Т.е. использовать надо `e.Message`.

CodingFeles commented on task-2/Summator/Summator/Program.cs:
```diff
@@ -10,24 +10,15 @@ static void Main(string[] args)
-                Console.WriteLine("Введите числовые значения.");
-            }
-            finally
+            catch (FormatException)
```
А хватит ли нам только FormatException? @Tauders ? Я бы просто базовый Exception использовал.

Tauders replied:
Я бы тоже базовый использовал

CodingFeles reviewed (changes requested):
Исправил большую часть как надо. Осталось немножко дожать обработку исключений.

St1ke committed:
[task-2] Обработал исключение типа Exception
